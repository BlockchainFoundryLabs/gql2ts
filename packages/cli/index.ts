#!/usr/bin/env node
'use strict';
import * as program from 'commander';
import {
  generateNamespace,
  writeNamespaceToFile,
  readFile,
  PossibleSchemaInput
} from '@gql2ts/util';
import { schemaToInterfaces, IOptions } from '@gql2ts/from-schema';

program
  .version('1.0.0')
  .usage('[options] <schema.json>')
  .option('-o --output-file [outputFile]', 'name for output file, will use stdout if not specified')
  .option('-n --namespace [namespace]', 'name for the namespace, defaults to "GQL"', 'GQL')
  .option('-i --ignored-types <ignoredTypes>', 'names of types to ignore (comma delimited)', v => v.split(','), [])
  .option('-l --legacy', 'Use TypeScript 1.x annotation', false)
  .parse(process.argv);

const run: (schema: PossibleSchemaInput, options: Partial<IOptions>) => void = (schema, options) => {
  let interfaces: string = schemaToInterfaces(schema, options);

  let namespace: string = generateNamespace(options.namespace!, interfaces);

  if (options.outputFile) {
    writeNamespaceToFile(options.outputFile, namespace);
  } else {
    console.log(namespace);
  }
};

const fileName: string | undefined = program.args[0];

if (!process.stdin.isTTY) {
  let input: string = '';
  process.stdin.resume();
  process.stdin.setEncoding('utf8');
  process.stdin.on('data', (data) => {
    input += data;
  });
  process.stdin.on('end', () => run(JSON.parse(input), program));
} else if (fileName) {
  const schema: string = readFile(fileName);
  run(schema, program);
} else {
  console.error('No input specified. Please use stdin or a file name.');
}

if (!process.argv.slice(2).length) {
  program.outputHelp();
}
