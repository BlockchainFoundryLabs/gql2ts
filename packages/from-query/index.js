"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const util_1 = require("@gql2ts/util");
const DEFAULT_FORMAT_INTERFACE = (opName, fields) => `export interface ${opName} {
${fields.join('\n  ')}
}`;
const DEFAULT_FORMAT_VARIABLES = (opName, fields) => `export interface ${opName}Input {
  ${fields.join('\n  ')}
}`;
const DEFAULT_FORMAT_FRAGMENT = (opName, fields, ext) => `export interface ${opName}${ext} {
${fields.join('\n')}
}`;
const DEFAULT_BUILD_ROOT_INTERFACE_NAME = def => {
    if (def.kind === 'OperationDefinition') {
        return def.name ? def.name.value : 'Anonymous';
    }
    else if (def.kind === 'FragmentDefinition') {
        return `IFragment${def.name.value}`;
    }
    else {
        throw new Error(`Unsupported Definition ${def.kind}`);
    }
};
const DEFAULT_TYPE_MAP = {
    ID: 'string',
    String: 'string',
    Boolean: 'boolean',
    Float: 'number',
    Int: 'number',
};
const DEFAULT_WRAP_LIST = type => `Array<${type}>`;
const DEFAULT_WRAP_PARTIAL = possiblePartial => {
    if (possiblePartial.isPartial) {
        return `Partial<${possiblePartial.iface}>`;
    }
    else {
        return possiblePartial.iface;
    }
};
const DEFAULT_GENERATE_SUBTYPE_INTERFACE_NAME = (selectionName, generatedCount) => `SelectionOn${selectionName}${!!generatedCount ? generatedCount : ''}`;
;
const DEFAULT_OPTIONS = {
    buildRootInterfaceName: DEFAULT_BUILD_ROOT_INTERFACE_NAME,
    formatVariableInterface: DEFAULT_FORMAT_VARIABLES,
    formatInterface: DEFAULT_FORMAT_INTERFACE,
    formatFragmentInterface: DEFAULT_FORMAT_FRAGMENT,
    wrapList: DEFAULT_WRAP_LIST,
    wrapPartial: DEFAULT_WRAP_PARTIAL,
    generateSubTypeInterfaceName: DEFAULT_GENERATE_SUBTYPE_INTERFACE_NAME,
};
;
const doIt = (schema, query, typeMap = {}, providedOptions = {}) => {
    const TypeMap = Object.assign({}, DEFAULT_TYPE_MAP, typeMap);
    const options = Object.assign({}, DEFAULT_OPTIONS, providedOptions);
    const { buildRootInterfaceName, formatFragmentInterface, formatInterface, formatVariableInterface, wrapList, wrapPartial, generateSubTypeInterfaceName, } = options;
    const parsedSchema = util_1.schemaFromInputs(schema);
    const parsedSelection = graphql_1.parse(query);
    function isNonNullable(type) {
        return type instanceof graphql_1.GraphQLNonNull;
    }
    function isList(type) {
        return type instanceof graphql_1.GraphQLList;
    }
    const printType = (type, isNonNull) => isNonNull ? type : `${type} | null`;
    const handleInputObject = (type, isNonNull) => {
        const variables = Object.keys(type.getFields()).map(k => type.getFields()[k]);
        const builder = `{\n${variables.map(v => `    ${v.name}?: ${convertToType(v.type)};`).join('\n')}\n  }`;
        return printType(builder, isNonNull);
    };
    const handleEnum = (type, isNonNull) => {
        const decl = type.getValues().map(en => `'${en.value}'`).join(' | ');
        return printType(decl, isNonNull);
    };
    const handleNamedTypeInput = (type, isNonNull) => {
        if (type.kind === 'NamedType' && type.name.kind === 'Name' && type.name.value) {
            const newType = parsedSchema.getType(type.name.value);
            if (newType instanceof graphql_1.GraphQLEnumType) {
                return handleEnum(newType, isNonNull);
            }
            else if (newType instanceof graphql_1.GraphQLInputObjectType) {
                return handleInputObject(newType, isNonNull);
            }
        }
    };
    const handleRegularType = (type, isNonNull, replacement) => {
        const typeValue = (typeof type.name === 'string') ? type.toString() : type.name.value;
        const showValue = replacement ? replacement : typeValue;
        const show = TypeMap[showValue] || (replacement ? showValue : 'any');
        return printType(show, isNonNull);
    };
    const convertVariable = (type, isNonNull = false, replacement = null) => {
        if (type.kind === 'ListType') {
            return wrapList(convertVariable(type.type, false, replacement)) + printType('', isNonNull);
        }
        else if (type.kind === 'NonNullType') {
            return convertVariable(type.type, true, replacement);
        }
        else {
            return handleNamedTypeInput(type, isNonNull) || handleRegularType(type, isNonNull, replacement);
        }
    };
    const convertToType = (type, isNonNull = false, replacement = null) => {
        if (isList(type)) {
            return wrapList(convertToType(type.ofType, false, replacement)) + printType('', isNonNull);
        }
        else if (isNonNullable(type)) {
            return convertToType(type.ofType, true, replacement);
        }
        else if (type instanceof graphql_1.GraphQLEnumType) {
            return handleEnum(type, isNonNull);
        }
        else {
            return handleRegularType(type, isNonNull, replacement);
        }
    };
    const UndefinedDirectives = new Set(['include', 'skip']);
    const isUndefinedFromDirective = directives => {
        if (!directives || !directives.length) {
            return false;
        }
        const badDirectives = directives.filter(d => !UndefinedDirectives.has(d.name.value));
        const hasDirectives = directives.some(d => UndefinedDirectives.has(d.name.value));
        if (badDirectives.length) {
            console.error('Found some unknown directives:');
            badDirectives.forEach(d => console.error(d.name.value));
        }
        if (hasDirectives) {
            return true;
        }
        else {
            return false;
        }
    };
    const getOperationFields = operation => {
        switch (operation) {
            case 'query':
                return parsedSchema.getQueryType();
            case 'mutation':
                return parsedSchema.getMutationType();
            case 'subscription':
                return parsedSchema.getSubscriptionType();
            default:
                throw new Error('Unsupported Operation');
        }
    };
    const getChildSelections = (operation, selection, indentation = '', parent, isUndefined = false) => {
        let str = '';
        let field;
        let isFragment = false;
        let isPartial = false;
        let generatedTypeCount = 0;
        let complexTypes = [];
        if (selection.kind === 'Field') {
            if (parent && graphql_1.isCompositeType(parent)) {
                if (parent instanceof graphql_1.GraphQLUnionType) {
                    field = parent.getTypes().map(t => t.getFields()[selection.name.value]).find(z => !!z);
                }
                else {
                    field = parent.getFields()[selection.name.value];
                }
            }
            else {
                const operationFields = getOperationFields(operation);
                field = operationFields.getFields()[selection.name.value];
            }
            let selectionName = selection.name.value;
            if (selection.alias) {
                selectionName = selection.alias.value;
            }
            if (isUndefined || isUndefinedFromDirective(selection.directives)) {
                selectionName += '?';
            }
            str += indentation + selectionName + ': ';
            if (!!selection.selectionSet) {
                let newParent;
                if (!field) {
                    console.log(selection, newParent);
                }
                const fieldType = graphql_1.getNamedType(field.type);
                if (graphql_1.isCompositeType(fieldType)) {
                    newParent = fieldType;
                }
                const selections = selection.selectionSet.selections.map(sel => getChildSelections(operation, sel, indentation + '  ', newParent));
                const nonFragments = selections.filter(s => !s.isFragment);
                const fragments = selections.filter(s => s.isFragment);
                const andOps = [];
                complexTypes.push(...selections.map(sel => sel.complexTypes).reduce((acc, arr) => { acc.push(...arr); return acc; }, []));
                if (nonFragments.length) {
                    const nonPartialNonFragments = nonFragments.filter(nf => !nf.isPartial);
                    const partialNonFragments = nonFragments.filter(nf => nf.isPartial);
                    if (nonPartialNonFragments.length) {
                        let builder = '';
                        builder += '{\n';
                        builder += nonPartialNonFragments.map(f => f.iface).join('\n');
                        builder += `\n${indentation}}`;
                        const newInterfaceName = generateSubTypeInterfaceName(selection.name.value, generatedTypeCount);
                        if (!newInterfaceName) {
                            andOps.push(builder);
                        }
                        else {
                            andOps.push(newInterfaceName);
                        }
                        generatedTypeCount += 1;
                        complexTypes.push({ iface: builder, isPartial: false, name: newInterfaceName });
                    }
                    if (partialNonFragments.length) {
                        let builder = '';
                        builder += 'Partial<{\n';
                        builder += partialNonFragments.map(f => f.iface).join('\n');
                        builder += `\n${indentation}}>`;
                        andOps.push(builder);
                        const newInterfaceName = generateSubTypeInterfaceName(selection.name.value, generatedTypeCount);
                        generatedTypeCount += 1;
                        complexTypes.push({ iface: builder, isPartial: true, name: newInterfaceName });
                    }
                }
                if (fragments.length) {
                    andOps.push(...fragments.map(wrapPartial));
                }
                const childType = andOps.join(' & ');
                str += convertToType(field.type, false, childType) + ';';
            }
            else {
                if (!field) {
                    console.log(selection);
                }
                str += convertToType(field.type) + ';';
            }
        }
        else if (selection.kind === 'FragmentSpread') {
            str = `IFragment${selection.name.value}`;
            isFragment = true;
            isPartial = isUndefinedFromDirective(selection.directives);
        }
        else if (selection.kind === 'InlineFragment') {
            const anon = !selection.typeCondition;
            if (!anon) {
                const typeName = selection.typeCondition.name.value;
                parent = parsedSchema.getType(typeName);
            }
            const selections = selection.selectionSet.selections.map(sel => getChildSelections(operation, sel, indentation, parent, !anon));
            let joinSelections = selections.map(s => s.iface).join('\n');
            isPartial = isUndefinedFromDirective(selection.directives);
            return {
                iface: joinSelections,
                isFragment,
                isPartial,
                complexTypes,
            };
        }
        else {
            throw new Error('Unsupported SelectionNode');
        }
        return {
            iface: str,
            isFragment,
            isPartial,
            complexTypes,
        };
    };
    const getVariables = variables => (variables.map(v => {
        const optional = v.type.kind !== 'NonNullType';
        return `${v.variable.name.value}${optional ? '?:' : ':'} ${convertVariable(v.type)};`;
    }));
    const variablesToInterface = (opName, variables) => {
        if (!variables || !variables.length) {
            return '';
        }
        const variableTypeDefs = getVariables(variables);
        return formatVariableInterface(opName, variableTypeDefs);
    };
    const buildAdditionalTypes = children => {
        const subTypes = children.reduce((acc, child) => { acc.push(...child.complexTypes); return acc; }, []);
        return subTypes.map(subtype => {
            if (subtype.isPartial) {
                return `export type ${subtype.name} = ${subtype.iface};`;
            }
            else {
                return `export interface ${subtype.name} ${subtype.iface}`;
            }
        });
    };
    return parsedSelection.definitions.map(def => {
        const ifaceName = buildRootInterfaceName(def);
        if (def.kind === 'OperationDefinition') {
            const variableInterface = variablesToInterface(ifaceName, def.variableDefinitions);
            const ret = def.selectionSet.selections.map(sel => getChildSelections(def.operation, sel, '  '));
            const fields = ret.map(x => x.iface);
            const iface = formatInterface(ifaceName, fields);
            const additionalTypes = buildAdditionalTypes(ret);
            return {
                variables: variableInterface,
                interface: iface,
                additionalTypes,
            };
        }
        else if (def.kind === 'FragmentDefinition') {
            // get the correct type
            const onType = def.typeCondition.name.value;
            const foundType = parsedSchema.getType(onType);
            const ret = def.selectionSet.selections.map(sel => getChildSelections('query', sel, '  ', foundType));
            const ext = ret.filter(x => x.isFragment).map(x => x.iface).join(', ');
            const extensions = ext ? ` extends ${ext}` : '';
            const fields = ret.filter(x => !x.isFragment).map(x => x.iface);
            const iface = formatFragmentInterface(ifaceName, fields, extensions);
            const additionalTypes = buildAdditionalTypes(ret);
            return {
                interface: iface,
                variables: '',
                additionalTypes,
            };
        }
        else {
            throw new Error(`Unsupported Definition ${def.kind}`);
        }
    });
};
exports.default = doIt;
//# sourceMappingURL=index.js.map