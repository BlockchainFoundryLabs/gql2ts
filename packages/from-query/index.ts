import {
  parse,
  GraphQLSchema,
  DocumentNode,
  GraphQLField,
  GraphQLCompositeType,
  isCompositeType,
  getNamedType,
  GraphQLType,
  VariableDefinitionNode,
  TypeNode,
  OperationTypeNode,
  GraphQLObjectType,
  GraphQLEnumType,
  DirectiveNode,
  GraphQLInputObjectType,
  GraphQLInputField,
  GraphQLUnionType,
  GraphQLNamedType,
  FieldNode,
} from 'graphql';
import {
  schemaFromInputs,
  isList,
  isNonNullable,
  isEnum,
} from '@gql2ts/util';
import {
  ChildSelectionsType,
  IChildren,
  IComplexTypeSignature,
  IOptions,
  Signature,
  RegularTypeSignature,
  VariableTypeSignature,
  convertToTypeSignature,
  ITypeMap,
} from '@gql2ts/types';
import {
  DEFAULT_TYPE_MAP,
  DEFAULT_OPTIONS,
} from '@gql2ts/language-typescript';
import {
  GenerateSubtypeCache,
  SubtypeNamerAndDedupe,
  ISubtypeMetadata,
} from './subtype';

const doIt: Signature = (schema, query, typeMap= {}, providedOptions= {}) => {
  const TypeMap: ITypeMap = {
    ...DEFAULT_TYPE_MAP,
    ...typeMap
  };

  const {
    buildRootInterfaceName,
    formatFragmentInterface,
    formatInterface,
    formatVariableInterface,
    wrapList,
    wrapPartial,
    generateSubTypeInterfaceName,
    printType,
    formatInput,
    generateFragmentName,
    generateQueryName,
    formatEnum,
    defaultIndentation,
    interfaceBuilder,
    typeBuilder,
    typeJoiner,
    generateInterfaceDeclaration,
    exportFunction,
  }: IOptions = { ...DEFAULT_OPTIONS, ...providedOptions };

  const getSubtype: SubtypeNamerAndDedupe = GenerateSubtypeCache();

  const parsedSchema: GraphQLSchema = schemaFromInputs(schema);
  const parsedSelection: DocumentNode = parse(query);

  const handleInputObject: (type: GraphQLInputObjectType, isNonNull: boolean) => string = (type, isNonNull) => {
    const variables: GraphQLInputField[] = Object.keys(type.getFields()).map(k => type.getFields()[k]);
    const variableDeclarations: string[] = variables.map(v => formatInput(v.name, true, convertToType(v.type)));
    const builder: string = generateInterfaceDeclaration(variableDeclarations.map(v => `    ${v}`), defaultIndentation);
    return printType(builder, isNonNull);
  };

  const handleEnum: (type: GraphQLEnumType, isNonNull: boolean) => string = (type, isNonNull) => {
    const decl: string = formatEnum(type.getValues());
    return printType(decl, isNonNull);
  };

  const handleNamedTypeInput: (type: TypeNode, isNonNull: boolean) => string | undefined = (type, isNonNull) => {
    if (type.kind === 'NamedType' && type.name.kind === 'Name' && type.name.value) {
      const newType: GraphQLType = parsedSchema.getType(type.name.value);
      if (newType instanceof GraphQLEnumType) {
        return handleEnum(newType, isNonNull);
      } else if (newType instanceof GraphQLInputObjectType) {
        return handleInputObject(newType, isNonNull);
      }
    }
  };

  const handleRegularType: RegularTypeSignature = (type, isNonNull, replacement) => {
    const typeValue: string = (typeof type.name === 'string') ? type.toString() : type.name.value;
    const showValue: string = replacement || typeValue;
    const show: string = TypeMap[showValue] || (replacement ? showValue : TypeMap.__DEFAULT);
    return printType(show, isNonNull);
  };

  const convertVariable: VariableTypeSignature = (type, isNonNull= false, replacement= null) => {
    if (type.kind === 'ListType') {
      return printType(wrapList(convertVariable(type.type, false, replacement)), isNonNull!);
    } else if (type.kind === 'NonNullType') {
      return convertVariable(type.type, true, replacement);
    } else {
      return handleNamedTypeInput(type, isNonNull!) || handleRegularType(type, isNonNull!, replacement!);
    }
  };

  const convertToType: convertToTypeSignature = (type, isNonNull= false, replacement= null): string => {
    if (isList(type)) {
      return printType(wrapList(convertToType(type.ofType, false, replacement)), isNonNull!);
    } else if (isNonNullable(type)) {
      return convertToType(type.ofType, true, replacement);
    } else if (isEnum(type)) {
      return handleEnum(type, isNonNull!);
    } else {
      return handleRegularType(type, isNonNull!, replacement!);
    }
  };

  const UndefinedDirectives: Set<string> = new Set(['include', 'skip']);

  const isUndefinedFromDirective: (directives: DirectiveNode[] | undefined) => boolean = directives => {
    if (!directives || !directives.length) { return false; }

    const badDirectives: DirectiveNode[] = directives.filter(d => !UndefinedDirectives.has(d.name.value));
    const hasDirectives: boolean = directives.some(d => UndefinedDirectives.has(d.name.value));

    if (badDirectives.length) {
      console.error('Found some unknown directives:');
      badDirectives.forEach(d => console.error(d.name.value));
    }

    return hasDirectives;
  };

  const getOperationFields: (operation: OperationTypeNode) => GraphQLObjectType = operation => {
    switch (operation) {
      case 'query':
        return parsedSchema.getQueryType();
      case 'mutation':
        return parsedSchema.getMutationType();
      case 'subscription':
        return parsedSchema.getSubscriptionType();
      default:
        throw new Error('Unsupported Operation');
    }
  };

  const wrapPossiblePartial: (possiblePartial: IChildren) => string = possiblePartial => {
    if (possiblePartial.isPartial) {
      return wrapPartial(possiblePartial.iface);
    } else {
      return possiblePartial.iface;
    }
  };

  const flattenComplexTypes: (children: IChildren[]) => IComplexTypeSignature[] = children => (
    children.reduce((acc, child) => { acc.push(...child.complexTypes); return acc; }, [] as IComplexTypeSignature[])
  );

  // @FIXME this is pretty bad
  const fixIndentationForSubtype: (decl: string) => string =
    decl => decl.replace(/\n\s+/g, `\n${defaultIndentation}`).replace(defaultIndentation + '}', '}');

  type GetField = (operation: OperationTypeNode, selection: FieldNode, parent?: GraphQLType) => GraphQLField<any, any>;
  const getField: GetField = (operation, selection, parent) => {
    if (parent && isCompositeType(parent)) {
      if (parent instanceof GraphQLUnionType) {
        return parent.getTypes().map(t => t.getFields()[selection.name.value]).find(z => !!z)!;
      } else {
        return parent.getFields()[selection.name.value];
      }
    } else {
      const operationFields: GraphQLObjectType = getOperationFields(operation);
      return operationFields.getFields()[selection.name.value];
    }
  };

  const getChildSelections: ChildSelectionsType = (operation, selection, indentation= '', parent?, isUndefined= false): IChildren => {
    let str: string = '';
    let isFragment: boolean = false;
    let isPartial: boolean = false;
    let complexTypes: IComplexTypeSignature[] = [];

    if (selection.kind === 'Field') {
      const field: GraphQLField<any, any> = getField(operation, selection, parent);
      const selectionName: string = selection.alias ? selection.alias.value : selection.name.value;
      let childType: string | undefined;

      isUndefined = isUndefined || isUndefinedFromDirective(selection.directives);
      let resolvedType: string;
      if (selectionName.startsWith('__')) {
        resolvedType = TypeMap.String;
      } else if (!!selection.selectionSet) {
        let newParent: GraphQLCompositeType | undefined;
        const fieldType: GraphQLNamedType = getNamedType(field.type);
        if (isCompositeType(fieldType)) {
          newParent = fieldType;
        }

        const selections: IChildren[] =
          selection.selectionSet.selections.map(sel => getChildSelections(operation, sel, indentation + defaultIndentation,  newParent));

        const nonFragments: IChildren[] = selections.filter(s => !s.isFragment);
        const fragments: IChildren[] = selections.filter(s => s.isFragment);
        const andOps: string[] = [];

        complexTypes.push(...flattenComplexTypes(selections));

        if (nonFragments.length) {
          const nonPartialNonFragments: IChildren[] = nonFragments.filter(nf => !nf.isPartial);
          const partialNonFragments: IChildren[] = nonFragments.filter(nf => nf.isPartial);

          if (nonPartialNonFragments.length) {
            const interfaceDeclaration: string = generateInterfaceDeclaration(nonPartialNonFragments.map(f => f.iface), indentation);
            const subtypeInfo: ISubtypeMetadata | null = getSubtype(selection, interfaceDeclaration, generateSubTypeInterfaceName);
            const newInterfaceName: string | null = subtypeInfo ? subtypeInfo.name : null;
            andOps.push(newInterfaceName || interfaceDeclaration);
            if (newInterfaceName && subtypeInfo && !subtypeInfo.dupe) {
              complexTypes.push({ iface: fixIndentationForSubtype(interfaceDeclaration), isPartial: false, name: newInterfaceName });
            }
          }

          if (partialNonFragments.length) {
            const interfaceDeclaration: string =
              wrapPartial(generateInterfaceDeclaration(partialNonFragments.map(f => f.iface), indentation));
            const subtypeInfo: ISubtypeMetadata | null = getSubtype(selection, interfaceDeclaration, generateSubTypeInterfaceName);
            const newInterfaceName: string | null = subtypeInfo ? subtypeInfo.name : null;
            andOps.push(newInterfaceName || interfaceDeclaration);
            if (newInterfaceName && subtypeInfo && !subtypeInfo.dupe) {
              complexTypes.push({ iface: fixIndentationForSubtype(interfaceDeclaration), isPartial: true, name: newInterfaceName });
            }
          }
        }

        andOps.push(...fragments.map(wrapPossiblePartial));
        childType = typeJoiner(andOps);
        resolvedType = convertToType(field.type, false, childType);
      } else {
        resolvedType = convertToType(field.type, false, childType);
      }
      str = formatInput(indentation + selectionName, isUndefined, resolvedType);
    } else if (selection.kind === 'FragmentSpread') {
      str = generateFragmentName(selection.name.value);
      isFragment = true;
      isPartial = isUndefinedFromDirective(selection.directives);
    } else if (selection.kind === 'InlineFragment') {
      const anon: boolean = !selection.typeCondition;
      if (!anon) {
        const typeName: string = selection.typeCondition!.name.value;
        parent = parsedSchema.getType(typeName);
      }

      const selections: IChildren[] =
        selection.selectionSet.selections.map(sel => getChildSelections(operation, sel, indentation, parent, !anon));

      let joinSelections: string = selections.map(s => s.iface).join('\n');
      isPartial = isUndefinedFromDirective(selection.directives);
      complexTypes.push(...flattenComplexTypes(selections));
      return {
        iface: joinSelections,
        isFragment,
        isPartial,
        complexTypes,
      };
    }

    return {
      iface: str,
      isFragment,
      isPartial,
      complexTypes,
    };
  };

  const getVariables: (variables: VariableDefinitionNode[]) => string[] = variables => (
    variables.map(v => {
      const optional: boolean = v.type.kind !== 'NonNullType';
      return formatInput(v.variable.name.value, optional, convertVariable(v.type));
    })
  );

  const variablesToInterface: (operationName: string, variables: VariableDefinitionNode[] | undefined) => string = (opName, variables) => {
    if (!variables || !variables.length) { return ''; }
    const variableTypeDefs: string[] = getVariables(variables);
    return exportFunction(formatVariableInterface(opName, variableTypeDefs));
  };

  const buildAdditionalTypes: (children: IChildren[]) => string[] = children => {
    const subTypes: IComplexTypeSignature[] = flattenComplexTypes(children);

    return subTypes.map(subtype => {
      if (subtype.isPartial) {
        return exportFunction(typeBuilder(subtype.name, subtype.iface));
      } else {
        return exportFunction(interfaceBuilder(subtype.name, subtype.iface));
      }
    });
  };

  return parsedSelection.definitions.map(def => {
    const ifaceName: string = buildRootInterfaceName(def, generateQueryName, generateFragmentName);
    if (def.kind === 'OperationDefinition') {
      const variableInterface: string = variablesToInterface(ifaceName, def.variableDefinitions);
      const ret: IChildren[] = def.selectionSet.selections.map(sel => getChildSelections(def.operation, sel, defaultIndentation));
      const fields: string[] = ret.map(x => x.iface);
      const iface: string = exportFunction(formatInterface(ifaceName, fields));
      const additionalTypes: string[] = buildAdditionalTypes(ret);

      return {
        variables: variableInterface,
        interface: iface,
        additionalTypes,
      };
    } else if (def.kind === 'FragmentDefinition') {
      // get the correct type
      const onType: string = def.typeCondition.name.value;
      const foundType: GraphQLType = parsedSchema.getType(onType);

      const ret: IChildren[] = def.selectionSet.selections.map(sel => getChildSelections('query', sel, defaultIndentation, foundType));
      const extensions: string[] = ret.filter(x => x.isFragment).map(x => x.iface);
      const fields: string[] = ret.filter(x => !x.isFragment).map(x => x.iface);
      const iface: string = exportFunction(formatFragmentInterface(ifaceName, fields, extensions));
      const additionalTypes: string[] = buildAdditionalTypes(ret);

      return {
        interface: iface,
        variables: '',
        additionalTypes,
      };
    } else {
      throw new Error(`Unsupported Definition ${def.kind}`);
    }
  });
};

export default doIt;
