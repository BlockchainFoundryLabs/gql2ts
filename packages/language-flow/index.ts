import { IFromQueryOptions, WrapType } from '@gql2ts/types';
import { DEFAULT_OPTIONS as TS_OPTIONS } from '@gql2ts/language-typescript';

export const FLOW_WRAP_PARTIAL: WrapType = partial => `$SHAPE<${partial}>`;

export const DEFAULT_OPTIONS: IFromQueryOptions = {
  ...TS_OPTIONS,
  wrapPartial: FLOW_WRAP_PARTIAL
};
export default DEFAULT_OPTIONS;
