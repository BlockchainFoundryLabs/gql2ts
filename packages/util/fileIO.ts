import * as fs from 'fs';

export const readFile: (fileName: string) => string = fileName => JSON.parse(fs.readFileSync(fileName).toString());

export const writeToFile: (fileName: string, data: string) => void =
  (fileName, data) => fs.writeFile(fileName, data, err => err && console.error('Failed to write', err) );
