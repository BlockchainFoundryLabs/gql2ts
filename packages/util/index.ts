export { readFile, writeToFile } from './fileIO';
export { generateNamespace, writeNamespaceToFile } from './namespace';

export {
  PossibleIntrospectionInputs,
  PossibleSchemaInput,
  isIntrospectionResult,
  schemaFromInputs
} from './schema';
