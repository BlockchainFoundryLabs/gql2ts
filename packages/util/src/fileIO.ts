import * as fs from 'fs';

export const badWriteHandler: (err?: Error) => void = err => { if (err) { throw err; } };
export const readFile: (fileName: string) => string = fileName => JSON.parse(fs.readFileSync(fileName).toString());
export const writeToFile: (fileName: string, data: string) => void =
  (fileName, data) => fs.writeFile(fileName, data, badWriteHandler);
