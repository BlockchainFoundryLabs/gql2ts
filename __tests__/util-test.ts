import * as fs from 'fs';
import * as utils from '../packages/util';
import { badWriteHandler } from '../packages/util/dist/fileIO';
import schemaLanguage from './shared/simpleSchema';
import { GraphQLSchema, introspectionQuery, graphql } from 'graphql';

const builtSchema: GraphQLSchema = utils.schemaFromInputs(schemaLanguage);

describe('schema', () => {
  it('works with schema language', () => {
    expect(builtSchema).toBeInstanceOf(GraphQLSchema);
  });

  describe('introspection query', () => {
    const promised: Promise<any> = graphql(builtSchema, introspectionQuery);

    it('works with introspected query in data key', () =>
      expect(
        promised.then(r => utils.schemaFromInputs(r))
      ).resolves.toBeInstanceOf(GraphQLSchema));

    it('works with introspected query in data key', () =>
      expect(
        promised.then(r => utils.schemaFromInputs(r.data))
      ).resolves.toBeInstanceOf(GraphQLSchema));
  });

  it('works with a schema', () => {
    expect(utils.schemaFromInputs(builtSchema)).toBeInstanceOf(GraphQLSchema);
  });

  it('throws on other', () => {
    expect(() => utils.schemaFromInputs({})).toThrowError(
      'Invalid Schema Input'
    );
  });
});

describe('IO stuff', () => {
  it('writes file', () => {
    // const writeFileStub = sinon.stub(fs, 'writeFile');
    const spy: jest.Spy = jest.spyOn(fs, 'writeFile').mockImplementation(() => null);
    utils.writeToFile('test', 'test');
    expect(spy).toHaveBeenCalledTimes(1);
    spy.mockReset();
    spy.mockRestore();
    spy.mockClear();
  });

  describe('error handler', () => {
    it('does nothing if called w/ undefined', () => {
      expect(() => badWriteHandler(undefined)).not.toThrow();
    });
    it('throws if called', () => {
      expect(() => badWriteHandler(err)).toThrowErrorMatchingSnapshot();
    });
  });
  it('writes file', () => {
    const spy: jest.Spy = jest.spyOn(fs, 'readFileSync').mockImplementation(() => '{}');
    utils.readFile('test');
    expect(spy).toHaveBeenCalledTimes(1);
    spy.mockReset();
    spy.mockRestore();
    spy.mockClear();
  });
});
