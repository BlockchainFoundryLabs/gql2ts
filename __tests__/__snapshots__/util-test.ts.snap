// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IO stuff error handler throws if called 1`] = `"Error test"`;
